File: 20/cpp/aoc5.cc
#include <algorithm>
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <math.h>
#include "aoc_helper.h"

int calculateByHalves(std::string seatString) {
    int upperRange = (int)pow(2, seatString.size()) - 1;
    int lowerRange = 0;

    for(int i = 0; i < seatString.size(); i++) {
        // std::cout << "Upper: " << upperRange << " Lower: " << lowerRange << std::endl;
        int diff = (upperRange - lowerRange) / 2 + 1;

        // std::cout << "Diff: " << diff << std::endl;

        char el = seatString[i];
        if (el == 'F' || el == 'L') {
            //take lower half
            upperRange = upperRange - diff;
        } else if (el == 'B' || el == 'R') {
            //take upper half
            lowerRange = lowerRange + diff;
        } else {
            std::cout << "Invalid input: " << el << std::endl;
            exit(1);
        }
    }
    // std::cout << "Upper: " << upperRange << " Lower: " << lowerRange << std::endl;
    return (upperRange + lowerRange) / 2;
}

int calculateSeatId(std::string seat) {
    int row = calculateByHalves(seat.substr(0, 7));
    int col = calculateByHalves(seat.substr(7, 3));

    return (row * 8) + col;
}

// part 1
// int highestSeat(std::vector<std::string> arr) {
//     int max = 0;
//     for (int i = 0; i < arr.size(); i++) {
//         std::string seat = arr[i];
//         int seatId = calculateSeatId(seat);
//         if (seatId > max) {
//             max = seatId;
//         }
//     }
//     return max;
// }

// part 2
void mySeat(std::vector<std::string> arr) {
    std::vector<int> seats;
    for (int i = 0; i < arr.size(); i++) {
        std::string seat = arr[i];
        int seatId = calculateSeatId(seat);
        seats.push_back(seatId);
    }
    std::sort(seats.begin(), seats.end());

    int lastSeat;
    int currentSeat;
    for(int i = 1; i < seats.size(); i++) {
        lastSeat = seats[i-1];
        currentSeat = seats[i];
        if (currentSeat - lastSeat != 1) {
            std::cout << "currentSeat " << currentSeat << " lastSeat " << lastSeat << std::endl;
        }
    }
}

int main() {
    //read input from file
    std::vector<std::string> data;
    data = getInputString("./input/aoc5.txt");

    mySeat(data);
}

File: 20/go/aoc5.go
package main

import "fmt"

func main() {
	data, err := ReadFile("../input/aoc5.txt")
	if err != nil {
		fmt.Printf("readFile failed: %v\n", err)
	}

	ans := highestSeatId(data)

	fmt.Printf("The highest seat ID is %d \n", ans)
}

func highestSeatId(data []string) int {
	highestSeatId := 0
	for _, seat := range data {
		if len(seat) < 1 {
			continue
		}
		seatId := seatId(seat)
		if seatId > highestSeatId {
			highestSeatId = seatId
		}
	}
	return highestSeatId
}

func seatId(seatCode string) int {
	// BBFFBBFRLL
	row := seatCode[:7]
	col := seatCode[7:]
	return seatBinary(row, "row")*8 + seatBinary(col, "col")
}

func seatBinary(seatCode string, seatType string) int {
	fmt.Println("seatCode: ", seatCode)
	lowerIdx := 0
	upperIdx := 7
	if seatType == "row" {
		upperIdx = 127
	}
	for i, c := range seatCode {
		diff := (upperIdx - lowerIdx + 1) / 2
		fmt.Println("upper: ", upperIdx)
		fmt.Println("lower: ", lowerIdx)
		switch c {
		case 'F', 'L':
			//take lower half
			if i < len(seatCode)-1 {
				upperIdx -= diff
			}
		case 'B', 'R':
			//take upper half
			if i < len(seatCode)-1 {
				lowerIdx += diff
			}
		}
	}
	fmt.Println("seatBinary: ", upperIdx+lowerIdx/2)
	return upperIdx + lowerIdx/2
}
